### ISSUE TITLE
False positives

### ISSUE THREAD

**Issue by robtesch** [icanhazstring/composer-unused#12]
Hi there,

Awesome idea for a package, however I am getting some false positives. In my case I am have a Laravel project, and the following packages came up as definite false positives:

`* aws/aws-sdk-php ✗
 * davejamesmiller/laravel-breadcrumbs ✗
 * doctrine/dbal ✗
 * jeremykenedy/laravel-phpinfo ✗
 * laravel/scout ✗
 * laravelcollective/html ✗`

There is one more package which I'm not 100% sure about: league/oauth2-client. I will investigate that one and get back to you.

Not 100% sure why these packages are hitting false positives, but might be to do with Laravel using a few of them internally, and others are being auto-discovered by Laravel, and not being used directly within the main app code.

**Comment by icanhazstring** [comment]
Hi @robtesch thanks for reaching out.

> Not 100% sure why these packages are hitting false positives, but might be to do with Laravel using a few of them internally, and others are being auto-discovered by Laravel, and not being used directly within the main app code.

This mostly happens, as you mentioned, if you require a package, which is mainly used by another required package. So for example:

If you require `doctrine/orm` AND `doctrine/dbal` without using any `doctrine/dbal` code directly in your code. This package will be marked as unused.

So normally you can safeley remove this `require` from your `composer.json`.
I could add another feature which might cross reference the package with another required package so you will get an output that you require a package which is used by another one but not in your code.

**Comment by Wharenn** [comment]
I hit similar false positive results with a Symfony project and it was a bit frightening 😄 

That cross check would be great! It would give user trust to remove direct requirements of these packages.

This package is a great idea!

**Comment by icanhazstring** [comment]
Feature idea: Cross reference used packages to inform user about required packages which are already installed via other 3rd party requirement. So these packages can be safely removed.

**Comment by hgraca** [comment]
I got quite a lot of false positives, for example `willdurand/js-translation-bundle`, which in my symfony project is only referenced in my `rounting.yml` and `bundles.yml`.

I guess my other false positives are also only referenced in configuration files.

**Comment by icanhazstring** [comment]
@hgraca hm that's quite a catch. And since symfony is used a lot. Maybe I should cover yml files as well.

**Comment by hkdobrev** [comment]
Symfony components has several cases where they would have a package relying on optional packages. The component would require these packages explicitly, but would use them if present or if you pass some arguments in a method. So you need to require these 3rd-party packages in your application as well if you want to use a component in a specific way. However, application code would not use them anywhere.

See https://github.com/symfony/symfony/issues/26259 for a specific example.

**Comment by icanhazstring** [comment]
So as is see it. These are packages listed in the composer `suggest` section. Which should also fall under the cross referencing. But in this case the usage should be states as **used** instead of **unused**

**Comment by hgraca** [comment]
@icanhazstring 
There are actually 3 files, including a node package that I guess is installed cozz of the PHP package.

`config/bundles.php`
```php
<?php

declare(strict_types=1);

return [
    // ...
    Bazinga\Bundle\JsTranslationBundle\BazingaJsTranslationBundle::class => ['all' => true],
    // ...
];
```

`config/routing.yaml`
```yaml
# ...
bazinga_js_translation:
    resource: "@BazingaJsTranslationBundle/Resources/config/routing/routing.yml"
# ...
```

`web/node_modules/bazinga-translator/package.json`
```json
{
  "name": "bazinga-translator",
  "version": "2.6.6",
  "description": "A pretty nice way to use your translation messages generated by the BazingaJsTranslationBundle in your JavaScript.",
  "keywords": [
    "translation",
    "symfony"
  ],
  "license": "MIT",
  "author": {
    "name": "William Durand",
    "email": "will+git@drnd.me"
  },
  "contributors": [
    {
      "name": "Bruno Sampaio",
      "email": "bens.sampaio@gmail.com"
    }
  ],
  "files": [
    "js/translator.js",
    "public/js/translator.min.js"
  ],
  "main": "js/translator.js",
  "devDependencies": {
    "uglify-js": "^2.7.3"
  },
  "scripts": {
    "build": "uglifyjs js/translator.js --comments --compress --mangle --output=public/js/translator.min.js"
  }
}
```

**Comment by icanhazstring** [comment]
@hgraca OK. So your problem is basically, that the `config/bundles.php` is not scanned.

I am working on a fix, that all php files are scanned. Instead of only those defined in your psr directive of your composer.json

**Comment by hgraca** [comment]
cool! Yeah, I think that will solve plenty of cases.

**Comment by simPod** [comment]
Found these as false positives:

- bentools/iterable-functions
- dragonmantank/cron-expression (fork https://github.com/simpod/cron-expression)
- nette/http
- nette/utils
- sebastian/diff
- zendframework/zend-diactoros

Also cases like symfony/validator as `framework.validation: { enable_annotations: true }` is set in yaml. (However I found out I wanted to remove it :P )

Then everything used just via yaml / xml / twig (symfony/asset). Great plugin tho!

**Comment by icanhazstring** [comment]
I just reworked the usage scan to use all `*.php` files in the project without `vendor/`
If you install `dev-master` you can give it a try.

**Comment by icanhazstring** [comment]
Just released version `0.3.0` to address this kind of issue.
If the problem persists please let me know.

Closing it for now :)

**Comment by simPod** [comment]
@icanhazstring still shows [those packages](https://github.com/icanhazstring/composer-unused/issues/12#issuecomment-476377109) as unused even tho they're used 🤔

**Comment by icanhazstring** [comment]
@simPod try running `composer unused -vvv`
This should generate a dump file which you can upload.

So we can see what files are scanned and why these packages still come up.

**Comment by icanhazstring** [comment]
This might have to do with #22 
Which I will address in the next release.

**Comment by robtesch** [comment]
Right, I've updated to the latest version, it's improved things but not 100%, still getting a few packages that are showing up as unused when they actually are used. Again, in this case I think it's because of the way Laravel "auto-discovers" certain packages.

I've decided to focus on one specific package (jeremykenedy/laravel-phpinfo) as a way of seeing what the issue could be.

For that package, all you do to use it is visit the route "/phpinfo" on the front end and it shows the relevant info. There is no direct mention of the class anywhere within the PHP files of the app, only a mention of this route.

In the "olden days" you used to have to "register" a Laravel package by adding it to config/app.php providers array (where you would explicitly add the class name of the package), but that's no longer needed. I imagine that might be why this particular package is showing up as a false positive.

**Comment by icanhazstring** [comment]
Tha js for the clarification @robtesch this seems like a bumper. An idea could be to have framework specific scans. But that lies in the future.

For know the `--excludePackage` parameter could be helpful here. Since you know it's used 😉

**Comment by robtesch** [comment]
@icanhazstring thanks, will do that. Any chance we could add something to our composer.json that would show composer-unused that we want to exclude these packages? I am thinking of the case of working in a team, and somebody else runs composer-unused, sees a bunch of packages "unused" and then decided to chop them, without realising that they are, in fact, used.

**Comment by icanhazstring** [comment]
Maybe behind the `extras` directive could be possible. Sure. File a feature request :)

**Comment by icanhazstring** [comment]
I already created a new feature request #26 to ignore packages by definition. This should solve your problem @robtesch.

**Comment by robtesch** [comment]
@icanhazstring fantastic, thank you!

**Comment by hgraca** [comment]
The good news is that the package is not showing anymore as a false positive.

The bad news is that others still are:
```
 Unused packages
 * beberlei/doctrineextensions ✗ (config/config.yaml as `DoctrineExtensions\Query\Mysql\Regexp`)
 * overblog/dataloader-bundle ✗  (config/bundles.php)
 * perftools/xhgui-collector ✗ (web/app.php as `require_once __DIR__ . '/../vendor/perftools/xhgui-collector/external/header.php';`)
 * symfony/monolog-bundle ✗ (config/bundles.php)
```
The list gave more, but I need time to check if they are actually used or not.

**Comment by icanhazstring** [comment]
@hgraca thanks for more information.
The `*.yaml` file is noted. This is something I could add in the future.

Can you show me content of your `config/bundles.php` where the two packages are used?
Also can you file these informations in a new issue (false positives #2) so i can address them there.
Thank you 😉

**Comment by hgraca** [comment]
:+1: 
done.

**Comment by icanhazstring** [comment]
> See [symfony/symfony#26259](https://github.com/symfony/symfony/issues/26259) for a specific example.

Thank you @hkdobrev for pointing this out. Solved with #23
